<h1>Loading Necessary Libraries</h1>

In [68]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

<h2>Data PreProcessing Starts</h2>

In [69]:
df=pd.read_csv('test.csv')
df_survived=pd.read_csv('gender_submission.csv')

In [70]:
df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [71]:
df_survived.head()

PassengerId  Survived
0          892         0
1          893         1
2          894         0
3          895         0
4          896         1

# Concating df and df_survived because df doesnot have survived column and both are aligned perfectly


In [72]:
df=pd.concat([df,df_survived['Survived']],axis=1)

In [73]:
df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  Survived  
0  34.5      0      0   330911   7.8292   NaN        Q         0  
1  47.0      1      0   363272   7.0000   NaN        S         1  
2  62.0      0      0   240276   9.6875   NaN        Q         0  
3  27.0      0      0   315154   8.6625   NaN        S         0  
4  22.0      1      1  3101298  12.2875   NaN        S         1

In [74]:
df.sample(5)

PassengerId  Pclass                            Name     Sex   Age  SibSp  \
288         1180       3         Mardirosian, Mr. Sarkis    male   NaN      0   
72           964       3  Nieminen, Miss. Manta Josefina  female  29.0      0   
62           954       3    Bjorklund, Mr. Ernst Herbert    male  18.0      0   
401         1293       2                 Gale, Mr. Harry    male  38.0      1   
322         1214       2              Nesson, Mr. Israel    male  26.0      0   

     Parch   Ticket     Fare  Cabin Embarked  Survived  
288      0     2655   7.2292  F E46        C         0  
72       0  3101297   7.9250    NaN        S         1  
62       0   347090   7.7500    NaN        S         0  
401      0    28664  21.0000    NaN        S         0  
322      0   244368  13.0000     F2        S         0

In [75]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
 11  Survived     418 non-null    int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


In [76]:
df.shape

(418, 12)

In [ ]:
df.info()

# We have 418 total datas and in some we can observe value for nulls. 
Checking for null values for each columns

In [77]:
df.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
Survived         0
dtype: int64

# Determining how to fill datas for age


In [78]:
print('Oldest Passenger was of:',df['Age'].max(),'Years')
print('Youngest Passenger was of:',df['Age'].min(),'Years')
print('Average Age on the ship:',df['Age'].mean(),'Years')

#So maybe mean would be appropriate

Oldest Passenger was of: 76.0 Years
Youngest Passenger was of: 0.17 Years
Average Age on the ship: 30.272590361445783 Years



# Determining how to fill datas for Cabins


In [79]:
print(len(df['Cabin'].unique()))

#Many unique Values. 

print(df['Cabin'].sample(10))

#Values seem not appropiate to take mean.


77
414    C105
147     NaN
225     NaN
126     NaN
396     NaN
9       NaN
21      NaN
366     NaN
45      NaN
206     NaN
Name: Cabin, dtype: object


In [12]:

# Checking Cabin's Relevance to output


In [80]:

#df['Cabin'].corr(df['Survived']) Couldnot be computed

#Also if we analyze Cabin will have no importance to survive rate

#Best to drop them.

# Determining how to fill datas for Fare


In [81]:
print(len(df['Fare'].unique()))

#Has 170 unique Values. 
#So a Mean would be appropiate


170


## Lets Fill NAN Feilds

In [82]:
df['Fare'].fillna(df['Fare'].mean(),inplace=True) #Replacing with mean values
df['Age'].fillna(df['Age'].mean(),inplace=True) #Replacing with mean values
df.drop('Cabin',axis=1,inplace=True)

In [83]:
df.columns

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Embarked', 'Survived'],
      dtype='object')

Age has 177 null values
Cabin has 687 null values
Emabrked has 2 null values

We can drop them. But 177 and 687 are too large numbers to drop.
Further preprocessing might be required.

<h1> Now Lets Look at Feature Selection
 </h1>

In [84]:
#Looking at correlation between columns with survival

df_comp=df.iloc[:,df.columns!='Survived']
df_comp.corrwith(df['Survived'])


PassengerId   -0.023245
Pclass        -0.108615
Age           -0.000011
SibSp          0.099943
Parch          0.159120
Fare           0.191382
dtype: float64

In [85]:
#dropping PassengerID , Age , SibSp 
rows=['PassengerId','Age','SibSp']
for r in rows:
    df.drop(r,axis=1,inplace=True)

In [86]:
df.columns

Index(['Pclass', 'Name', 'Sex', 'Parch', 'Ticket', 'Fare', 'Embarked',
       'Survived'],
      dtype='object')

# Checking realations for<br>
## <b>Name  , Ticket And  Sex</b>


#We cannont find Correlation between Name and Survived

#Name Cannot be Categorized 

#Possibly Survival rate cannot be determined by only names.

#So better drop it


In [87]:
df.drop('Name',axis=1,inplace=True)

In [88]:
#Tickets

df['Ticket'].sample(10)

33            W./C. 6607
258                28404
167           W./C. 6608
189                28221
278             CA 31352
10                349220
151               349255
239             PC 17761
154               347077
97     STON/O 2. 3101268
Name: Ticket, dtype: object

# Tickets some have String values and they all seem to be random.
#So lets drop it as ticket.


In [89]:
df.drop('Ticket',axis=1,inplace=True)


In [90]:
# For Sex
df['Sex'].unique()

array(['male', 'female'], dtype=object)

# Lets map Male and Female to 0s and 1s so as to compare correlation


In [91]:
df['Sex']=df['Sex'].apply(lambda x:1 if x=='female' else 0)

# Lets Check correlation


In [92]:
df['Sex'].value_counts()
df['Sex'].corr(df['Survived'])

0.9999999999999999

Thats very high Correlation. So lets keep Sex as feature


In [93]:

df.columns.values

array(['Pclass', 'Sex', 'Parch', 'Fare', 'Embarked', 'Survived'],
      dtype=object)

# column names


In [94]:
col_names=df.columns.values 

for i in range(0,len(col_names)):
    uniq_datas=df[col_names[i]].unique() #unique datas of a column 
    
    if(len( uniq_datas ) < 10 ):
        print(col_names[i] ," : ", np.sort(uniq_datas))
    
    else:
        arr= df[col_names[i]]
        chunks = len(df[col_names[i]])/10
        print( col_names[i]," : " ,np.array_split(arr,chunks)) #splitting in chunks
    

Pclass  :  [1 2 3]
Sex  :  [0 1]
Parch  :  [0 1 2 3 4 5 6 9]
Fare  :  [0      7.8292
1      7.0000
2      9.6875
3      8.6625
4     12.2875
5      9.2250
6      7.6292
7     29.0000
8      7.2292
9     24.1500
10     7.8958
Name: Fare, dtype: float64, 11    26.0000
12    82.2667
13    26.0000
14    61.1750
15    27.7208
16    12.3500
17     7.2250
18     7.9250
19     7.2250
20    59.4000
21     3.1708
Name: Fare, dtype: float64, 22     31.6833
23     61.3792
24    262.3750
25     14.5000
26     61.9792
27      7.2250
28     30.5000
29     21.6792
30     26.0000
31     31.5000
32     20.5750
Name: Fare, dtype: float64, 33    23.4500
34    57.7500
35     7.2292
36     8.0500
37     8.6625
38     9.5000
39    56.4958
40    13.4167
41    26.5500
42     7.8500
43    13.0000
Name: Fare, dtype: float64, 44     52.5542
45      7.9250
46     29.7000
47      7.7500
48     76.2917
49     15.9000
50     60.0000
51     15.0333
52     23.0000
53    263.0000
54     15.5792
Name: Fare, dtype: float6

# Now converting Categorical data Embarked into codes


In [95]:

embarked_df=pd.get_dummies(df['Embarked'])

df=pd.concat([df,embarked_df],axis=1)

df.drop('Embarked',axis=1,inplace=True)


# Now converting Categorical data Parch into codes


In [96]:
parched_df=pd.get_dummies(df['Parch'])

df=pd.concat([df,parched_df],axis=1)

df.drop('Parch',axis=1,inplace=True)


In [97]:
df.head()

Pclass  Sex     Fare  Survived  C  Q  S  0  1  2  3  4  5  6  9
0       3    0   7.8292         0  0  1  0  1  0  0  0  0  0  0  0
1       3    1   7.0000         1  0  0  1  1  0  0  0  0  0  0  0
2       2    0   9.6875         0  0  1  0  1  0  0  0  0  0  0  0
3       3    0   8.6625         0  0  0  1  1  0  0  0  0  0  0  0
4       3    1  12.2875         1  0  0  1  0  1  0  0  0  0  0  0

# Features Cannot be Int so Changing The Feature Names

In [98]:
for i in range(0,10):
    df.rename(columns={i:f"Parch_{i}"},inplace=True)

In [99]:
df.head()

Pclass  Sex     Fare  Survived  C  Q  S  Parch_0  Parch_1  Parch_2  \
0       3    0   7.8292         0  0  1  0        1        0        0   
1       3    1   7.0000         1  0  0  1        1        0        0   
2       2    0   9.6875         0  0  1  0        1        0        0   
3       3    0   8.6625         0  0  0  1        1        0        0   
4       3    1  12.2875         1  0  0  1        0        1        0   

   Parch_3  Parch_4  Parch_5  Parch_6  Parch_9  
0        0        0        0        0        0  
1        0        0        0        0        0  
2        0        0        0        0        0  
3        0        0        0        0        0  
4        0        0        0        0        0

In [100]:
#Checking correlations
df.corr()

Pclass       Sex      Fare  Survived         C         Q  \
Pclass    1.000000 -0.108615 -0.576619 -0.108615 -0.318543  0.252608   
Sex      -0.108615  1.000000  0.191382  1.000000  0.033684  0.115574   
Fare     -0.576619  0.191382  1.000000  0.191382  0.312040 -0.155539   
Survived -0.108615  1.000000  0.191382  1.000000  0.033684  0.115574   
C        -0.318543  0.033684  0.312040  0.033684  1.000000 -0.199786   
Q         0.252608  0.115574 -0.155539  0.115574 -0.199786  1.000000   
S         0.120783 -0.105883 -0.178464 -0.105883 -0.767375 -0.474962   
Parch_0   0.054258 -0.224158 -0.256771 -0.224158 -0.080877  0.171098   
Parch_1  -0.093179  0.136987  0.148275  0.136987  0.140263 -0.109384   
Parch_2   0.034153  0.147542  0.096940  0.147542 -0.042399 -0.102952   
Parch_3  -0.094264  0.053559  0.232155  0.053559  0.083666 -0.029898   
Parch_4  -0.021898  0.091725  0.135019  0.091725 -0.039394 -0.024382   
Parch_5   0.042775 -0.037018 -0.003723 -0.037018 -0.027822 -0.017220   
Parch_6   0.042775 -0.037018  0.009898 -0.037018 -0.027822 -0.017220   
Parch_9   0.060565  0.019655  0.042173  0.019655 -0.039394 -0.024382   

                 S   Parch_0   Parch_1   Parch_2   Parch_3   Parch_4  \
Pclass    0.120783  0.054258 -0.093179  0.034153 -0.094264 -0.021898   
Sex      -0.105883 -0.224158  0.136987  0.147542  0.053559  0.091725   
Fare     -0.178464 -0.256771  0.148275  0.096940  0.232155  0.135019   
Survived -0.105883 -0.224158  0.136987  0.147542  0.053559  0.091725   
C        -0.767375 -0.080877  0.140263 -0.042399  0.083666 -0.039394   
Q        -0.474962  0.171098 -0.109384 -0.102952 -0.029898 -0.024382   
S         1.000000 -0.039328 -0.054392  0.105450 -0.055576  0.051335   
Parch_0  -0.039328  1.000000 -0.699793 -0.543545 -0.157850 -0.128729   
Parch_1  -0.054392 -0.699793  1.000000 -0.110354 -0.032048 -0.026135   
Parch_2   0.105450 -0.543545 -0.110354  1.000000 -0.024892 -0.020300   
Parch_3  -0.055576 -0.157850 -0.032048 -0.024892  1.000000 -0.005895   
Parch_4   0.051335 -0.128729 -0.026135 -0.020300 -0.005895  1.000000   
Parch_5   0.036256 -0.090916 -0.018458 -0.014337 -0.004164 -0.003395   
Parch_6   0.036256 -0.090916 -0.018458 -0.014337 -0.004164 -0.003395   
Parch_9   0.051335 -0.128729 -0.026135 -0.020300 -0.005895 -0.004808   

           Parch_5   Parch_6   Parch_9  
Pclass    0.042775  0.042775  0.060565  
Sex      -0.037018 -0.037018  0.019655  
Fare     -0.003723  0.009898  0.042173  
Survived -0.037018 -0.037018  0.019655  
C        -0.027822 -0.027822 -0.039394  
Q        -0.017220 -0.017220 -0.024382  
S         0.036256  0.036256  0.051335  
Parch_0  -0.090916 -0.090916 -0.128729  
Parch_1  -0.018458 -0.018458 -0.026135  
Parch_2  -0.014337 -0.014337 -0.020300  
Parch_3  -0.004164 -0.004164 -0.005895  
Parch_4  -0.003395 -0.003395 -0.004808  
Parch_5   1.000000 -0.002398 -0.003395  
Parch_6  -0.002398  1.000000 -0.003395  
Parch_9  -0.003395 -0.003395  1.000000

# Here Lets look between Pclass and Fare


In [101]:

df['Pclass'].corr(df['Fare'])

-0.5766190772528164

# A close correlation. Maybe we can discard a feature.
# checking between Survived


In [102]:

print("With Pclass")
print(df['Pclass'].corr(df['Survived']))

print("With Fare")
print(df['Fare'].corr(df['Survived']))

With Pclass
-0.10861452091523417
With Fare
0.19138247897510113


# Closer relation with Fare so lets drop PClass


In [103]:
df.drop('Pclass',axis=1,inplace=True)

In [104]:
df.head()

Sex     Fare  Survived  C  Q  S  Parch_0  Parch_1  Parch_2  Parch_3  \
0    0   7.8292         0  0  1  0        1        0        0        0   
1    1   7.0000         1  0  0  1        1        0        0        0   
2    0   9.6875         0  0  1  0        1        0        0        0   
3    0   8.6625         0  0  0  1        1        0        0        0   
4    1  12.2875         1  0  0  1        0        1        0        0   

   Parch_4  Parch_5  Parch_6  Parch_9  
0        0        0        0        0  
1        0        0        0        0  
2        0        0        0        0  
3        0        0        0        0  
4        0        0        0        0

<h1> Now Build Model </h1>

In [105]:
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
from sklearn.model_selection import GridSearchCV

In [106]:
#Now Finally Implementing Machine Learning
svc=SVC()
knn=KNeighborsClassifier()
dtc=DecisionTreeClassifier()
nb=GaussianNB()

In [107]:
models={'svc':svc,'knn':knn,'dtc':dtc,'nb':nb}

In [108]:
x,y=df.drop('Survived',axis=1),df['Survived']

In [127]:
train_x,test_x,train_y,test_y=train_test_split(x,y,test_size=0.2,random_state=42)

In [128]:
model_accuracy={}

for model in models.keys():
    train_x,test_x,train_y,test_y=train_test_split(x,y,test_size=0.2,random_state=500)
    mod = models[model].fit(train_x,train_y)
    predict=mod.predict(test_x)
    #print(f"{model}",classification_report(predict,test_y))
    print(f"{model}")
    print(confusion_matrix(predict,test_y))
    
    if model not in model_accuracy.keys():
        model_accuracy.update({model:accuracy_score(test_y,predict)})

svc
[[55 27]
 [ 1  1]]
knn
[[54  5]
 [ 2 23]]
dtc
[[56  0]
 [ 0 28]]
nb
[[56  0]
 [ 0 28]]


In [129]:
model_accuracy

{'svc': 0.6666666666666666, 'knn': 0.9166666666666666, 'dtc': 1.0, 'nb': 1.0}

# Lets try With GridSearch


In [130]:
params={
    'knn': {'n_neighbors':[3,5,7,9,11,13,15],
    'metric':['cosine','euclidean','manhattan'],
    'weights':['uniform','distance']},
    
    'svc': {
    'C':[0.1,1,10,100], 
    'gamma':[1,0.1,0.01,0.01], 
    'kernel':['rbf','linear']},
    
    'dtc':{
        'criterion':['gini','entropy'],
        'max_depth':[2,4,6,8,10,12]
    },
    
    'nb':{
        'priors': [None],
    'var_smoothing': [0.00000001, 0.000000001, 0.00000001]
    }
    
    }

In [131]:
model_accuracy={}
for model in models.keys():
    mod = GridSearchCV(
    models[model],
    params[model],
    verbose=1, #Progress bar showing
    cv=20, #cross validation
    n_jobs=-1, #cores to assign
)
    gridsearch_result=mod.fit(train_x,train_y)
    predict=mod.predict(test_x)
    print(f"{model} : ",gridsearch_result.best_estimator_)
    if model not in model_accuracy.keys():
        model_accuracy.update({model:accuracy_score(test_y,predict)})

Fitting 20 folds for each of 32 candidates, totalling 640 fits
svc :  SVC(C=0.1, gamma=1, kernel='linear')
Fitting 20 folds for each of 42 candidates, totalling 840 fits
knn :  KNeighborsClassifier(metric='cosine', n_neighbors=3, weights='distance')
Fitting 20 folds for each of 12 candidates, totalling 240 fits
dtc :  DecisionTreeClassifier(max_depth=2)
Fitting 20 folds for each of 3 candidates, totalling 60 fits
nb :  GaussianNB(var_smoothing=1e-08)


In [132]:
model_accuracy

{'svc': 1.0, 'knn': 0.9880952380952381, 'dtc': 1.0, 'nb': 1.0}

In [ ]:
#Maybe the result is nice.